// @polywrap/schema-parse/build/extract/env-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvVisitor = void 0;
var __1 = require("..");
var object_types_utils_1 = require("./utils/object-types-utils");
var visitorEnter = function (abi, state) { return ({
    ObjectTypeDefinition: function (node) {
        var typeName = node.name.value;
        if (__1.isEnvType(typeName)) {
            abi.envType = __1.createEnvDefinition({});
            state.currentType = abi.envType;
        }
    },
    NonNullType: function (_node) {
        state.nonNullType = true;
    },
    NamedType: function (node) {
        object_types_utils_1.extractNamedType(node, state);
    },
    ListType: function (_node) {
        object_types_utils_1.extractListType(state);
    },
    FieldDefinition: function (node) {
        object_types_utils_1.extractFieldDefinition(node, state);
    },
}); };
var visitorLeave = function (state) { return ({
    ObjectTypeDefinition: function (_node) {
        state.currentType = undefined;
    },
    FieldDefinition: function (_node) {
        state.currentProperty = undefined;
    },
    NonNullType: function (_node) {
        state.nonNullType = undefined;
    },
}); };
function getEnvVisitor(abi) {
    var state = {};
    return {
        enter: visitorEnter(abi, state),
        leave: visitorLeave(state),
    };
}
exports.getEnvVisitor = getEnvVisitor;
