// @polywrap/schema-parse/build/extract/imported-env-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getImportedEnvTypesVisitor = void 0;
var __1 = require("..");
var object_types_utils_1 = require("./utils/object-types-utils");
var imported_types_utils_1 = require("./utils/imported-types-utils");
var visitorEnter = function (importedEnvTypes, state) { return ({
    ObjectTypeDefinition: function (node) {
        var _a, _b;
        var imported = imported_types_utils_1.extractImportedDefinition(node, "env");
        if (!imported) {
            return;
        }
        var interfaces = (_a = node.interfaces) === null || _a === void 0 ? void 0 : _a.map(function (x) {
            return __1.createInterfaceImplementedDefinition({ type: x.name.value });
        });
        var importedType = __1.createImportedEnvDefinition({
            uri: imported.uri,
            namespace: imported.namespace,
            nativeType: imported.nativeType,
            interfaces: (interfaces === null || interfaces === void 0 ? void 0 : interfaces.length) ? interfaces : undefined,
            comment: (_b = node.description) === null || _b === void 0 ? void 0 : _b.value,
        });
        importedEnvTypes.push(importedType);
        state.currentType = importedType;
    },
    NonNullType: function (_node) {
        state.nonNullType = true;
    },
    NamedType: function (node) {
        object_types_utils_1.extractNamedType(node, state);
    },
    ListType: function (_node) {
        object_types_utils_1.extractListType(state);
    },
    FieldDefinition: function (node) {
        object_types_utils_1.extractFieldDefinition(node, state);
    },
}); };
var visitorLeave = function (state) { return ({
    ObjectTypeDefinition: function (_node) {
        state.currentType = undefined;
    },
    FieldDefinition: function (_node) {
        state.currentProperty = undefined;
    },
    NonNullType: function (_node) {
        state.nonNullType = undefined;
    },
}); };
var getImportedEnvTypesVisitor = function (abi) {
    var state = {};
    return {
        enter: visitorEnter(abi.importedEnvTypes || [], state),
        leave: visitorLeave(state),
    };
};
exports.getImportedEnvTypesVisitor = getImportedEnvTypesVisitor;
