// @polywrap/schema-parse/build/abi/utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scalarTypeNames = exports.isScalarType = exports.isImportedModuleType = exports.isModuleType = exports.MODULE_NAME = exports.isMapKeyType = exports.ScalarTypes = exports.MapKeyTypes = void 0;
exports.MapKeyTypes = {
    UInt: "UInt",
    UInt8: "UInt8",
    UInt16: "UInt16",
    UInt32: "UInt32",
    Int: "Int",
    Int8: "Int8",
    Int16: "Int16",
    Int32: "Int32",
    String: "String",
};
exports.ScalarTypes = __assign(__assign({}, exports.MapKeyTypes), { Boolean: "Boolean", Bytes: "Bytes", BigInt: "BigInt", BigNumber: "BigNumber", JSON: "JSON" });
function isMapKeyType(type) {
    return type in exports.MapKeyTypes;
}
exports.isMapKeyType = isMapKeyType;
exports.MODULE_NAME = "Module";
function isModuleType(type) {
    return type === exports.MODULE_NAME;
}
exports.isModuleType = isModuleType;
function isImportedModuleType(type) {
    return type.endsWith("_" + exports.MODULE_NAME);
}
exports.isImportedModuleType = isImportedModuleType;
function isScalarType(type) {
    return type in exports.ScalarTypes;
}
exports.isScalarType = isScalarType;
exports.scalarTypeNames = Object.keys(exports.ScalarTypes);
