// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.plugin/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePluginManifest = exports.deserializePluginManifest = exports.migratePluginManifest = exports.latestPluginManifestFormat = exports.PluginManifestSchemaFiles = exports.PluginManifestFormats = void 0;
var PluginManifestFormats;
(function (PluginManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    PluginManifestFormats["v0.1"] = "0.1";
    PluginManifestFormats["v0.1.0"] = "0.1.0";
    PluginManifestFormats["v0.2.0"] = "0.2.0";
})(PluginManifestFormats = exports.PluginManifestFormats || (exports.PluginManifestFormats = {}));
exports.PluginManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.plugin/0.1.0.json",
    "0.1.0": "formats/polywrap.plugin/0.1.0.json",
    "0.2.0": "formats/polywrap.plugin/0.2.0.json",
};
exports.latestPluginManifestFormat = PluginManifestFormats["v0.2.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migratePluginManifest", { enumerable: true, get: function () { return migrate_1.migratePluginManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializePluginManifest", { enumerable: true, get: function () { return deserialize_1.deserializePluginManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validatePluginManifest", { enumerable: true, get: function () { return validate_1.validatePluginManifest; } });
