// @polywrap/polywrap-manifest-types-js/build/formats/polywrap/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePolywrapManifest = exports.deserializePolywrapManifest = exports.migratePolywrapManifest = exports.latestPolywrapManifestFormat = exports.PolywrapManifestSchemaFiles = exports.PolywrapManifestFormats = void 0;
var PolywrapManifestFormats;
(function (PolywrapManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    PolywrapManifestFormats["v0.1"] = "0.1";
    PolywrapManifestFormats["v0.1.0"] = "0.1.0";
    PolywrapManifestFormats["v0.2.0"] = "0.2.0";
})(PolywrapManifestFormats = exports.PolywrapManifestFormats || (exports.PolywrapManifestFormats = {}));
exports.PolywrapManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap/0.1.0.json",
    "0.1.0": "formats/polywrap/0.1.0.json",
    "0.2.0": "formats/polywrap/0.2.0.json",
};
exports.latestPolywrapManifestFormat = PolywrapManifestFormats["v0.2.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migratePolywrapManifest", { enumerable: true, get: function () { return migrate_1.migratePolywrapManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializePolywrapManifest", { enumerable: true, get: function () { return deserialize_1.deserializePolywrapManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validatePolywrapManifest", { enumerable: true, get: function () { return validate_1.validatePolywrapManifest; } });
