// @polywrap/schema-bind/build/bindings/typescript/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isKeyword = exports.isBaseType = void 0;
var baseTypes = {
    boolean: "boolean",
    number: "number",
    string: "string",
};
function isBaseType(type) {
    return type in baseTypes;
}
exports.isBaseType = isBaseType;
// based on:
// - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#keywords
var keywords = {
    break: "break",
    case: "case",
    catch: "catch",
    class: "class",
    const: "const",
    continue: "continue",
    debugger: "debugger",
    default: "default",
    delete: "delete",
    do: "do",
    else: "else",
    export: "export",
    extends: "extends",
    false: "false",
    finally: "finally",
    for: "for",
    function: "function",
    if: "if",
    import: "import",
    in: "in",
    instanceof: "instanceof",
    new: "new",
    null: "null",
    return: "return",
    super: "super",
    switch: "switch",
    this: "this",
    throw: "throw",
    true: "true",
    try: "try",
    typeof: "typeof",
    var: "var",
    void: "void",
    while: "while",
    with: "with",
    yield: "yield",
    let: "let",
    await: "await",
    enum: "enum",
    implements: "implements",
    interface: "interface",
    package: "package",
    private: "private",
    protected: "protected",
    public: "public",
    static: "static",
    arguments: "arguments",
    eval: "eval",
};
function isKeyword(keyword) {
    return keyword in keywords;
}
exports.isKeyword = isKeyword;
