// @polywrap/ens-resolver-plugin-js/build/wrap/types.js
"use strict";
/// NOTE: This is an auto-generated file.
///       All modifications will be overwritten.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ethereum_Module = exports.UriResolver_Module = void 0;
/* URI: "ens/uri-resolver.core.polywrap.eth" */
exports.UriResolver_Module = {
    tryResolveUri: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/uri-resolver.core.polywrap.eth",
                    method: "tryResolveUri",
                    args: args
                })];
        });
    }); },
    getFile: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/uri-resolver.core.polywrap.eth",
                    method: "getFile",
                    args: args
                })];
        });
    }); }
};
/* URI: "ens/ethereum.polywrap.eth" */
exports.Ethereum_Module = {
    callContractView: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "callContractView",
                    args: args
                })];
        });
    }); },
    callContractStatic: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "callContractStatic",
                    args: args
                })];
        });
    }); },
    getBalance: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getBalance",
                    args: args
                })];
        });
    }); },
    encodeParams: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "encodeParams",
                    args: args
                })];
        });
    }); },
    encodeFunction: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "encodeFunction",
                    args: args
                })];
        });
    }); },
    solidityPack: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "solidityPack",
                    args: args
                })];
        });
    }); },
    solidityKeccak256: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "solidityKeccak256",
                    args: args
                })];
        });
    }); },
    soliditySha256: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "soliditySha256",
                    args: args
                })];
        });
    }); },
    getSignerAddress: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getSignerAddress",
                    args: args
                })];
        });
    }); },
    getSignerBalance: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getSignerBalance",
                    args: args
                })];
        });
    }); },
    getSignerTransactionCount: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getSignerTransactionCount",
                    args: args
                })];
        });
    }); },
    getGasPrice: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getGasPrice",
                    args: args
                })];
        });
    }); },
    estimateTransactionGas: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "estimateTransactionGas",
                    args: args
                })];
        });
    }); },
    estimateContractCallGas: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "estimateContractCallGas",
                    args: args
                })];
        });
    }); },
    checkAddress: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "checkAddress",
                    args: args
                })];
        });
    }); },
    toWei: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "toWei",
                    args: args
                })];
        });
    }); },
    toEth: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "toEth",
                    args: args
                })];
        });
    }); },
    awaitTransaction: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "awaitTransaction",
                    args: args
                })];
        });
    }); },
    waitForEvent: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "waitForEvent",
                    args: args
                })];
        });
    }); },
    getNetwork: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "getNetwork",
                    args: args
                })];
        });
    }); },
    requestAccounts: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "requestAccounts",
                    args: args
                })];
        });
    }); },
    callContractMethod: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "callContractMethod",
                    args: args
                })];
        });
    }); },
    callContractMethodAndWait: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "callContractMethodAndWait",
                    args: args
                })];
        });
    }); },
    sendTransaction: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "sendTransaction",
                    args: args
                })];
        });
    }); },
    sendTransactionAndWait: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "sendTransactionAndWait",
                    args: args
                })];
        });
    }); },
    deployContract: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "deployContract",
                    args: args
                })];
        });
    }); },
    signMessage: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "signMessage",
                    args: args
                })];
        });
    }); },
    sendRPC: function (args, client) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, client.invoke({
                    uri: "ens/ethereum.polywrap.eth",
                    method: "sendRPC",
                    args: args
                })];
        });
    }); }
};
/// Imported Modules END ///
