// polywrap/build/lib/helpers/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSimpleClient = void 0;
var ens_resolver_plugin_js_1 = require("@polywrap/ens-resolver-plugin-js");
var ethereum_plugin_js_1 = require("@polywrap/ethereum-plugin-js");
var ipfs_plugin_js_1 = require("@polywrap/ipfs-plugin-js");
var client_js_1 = require("@polywrap/client-js");
var client_config_builder_js_1 = require("@polywrap/client-config-builder-js");
function getSimpleClient(config) {
    var ensAddress = config.ensAddress, ethProvider = config.ethProvider, ipfsProvider = config.ipfsProvider;
    var plugins = [];
    var envs = [];
    if (ensAddress) {
        plugins.push({
            uri: "wrap://ens/ens-resolver.polywrap.eth",
            plugin: ens_resolver_plugin_js_1.ensResolverPlugin({
                addresses: {
                    testnet: ensAddress,
                },
            }),
        });
    }
    if (ethProvider) {
        plugins.push({
            uri: "wrap://ens/ethereum.polywrap.eth",
            plugin: ethereum_plugin_js_1.ethereumPlugin({
                connections: new ethereum_plugin_js_1.Connections({
                    networks: {
                        testnet: new ethereum_plugin_js_1.Connection({
                            provider: ethProvider,
                        }),
                    },
                }),
            }),
        });
    }
    if (ipfsProvider) {
        plugins.push({
            uri: "wrap://ens/ipfs.polywrap.eth",
            plugin: ipfs_plugin_js_1.ipfsPlugin({}),
        });
        envs.push({
            uri: "wrap://ens/ipfs.polywrap.eth",
            env: {
                provider: ipfsProvider,
                fallbackProviders: client_config_builder_js_1.defaultIpfsProviders,
            },
        });
    }
    return new client_js_1.PolywrapClient({ plugins: plugins });
}
exports.getSimpleClient = getSimpleClient;
