// @polywrap/wrap-manifest-types-js/build/formats/wrap.info/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateTypes.js to regenerate this file.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateWrapManifest = exports.serializeWrapManifest = exports.deserializeWrapManifest = exports.migrateWrapManifest = exports.latestWrapAbiVersion = exports.latestWrapManifestVersion = exports.WrapAbiVersions = exports.WrapManifestAbiVersions = exports.WrapManifestVersions = exports.WrapManifestSchema_0_1 = void 0;
__exportStar(require("./0.1"), exports);
var _0_1_schema_json_1 = __importDefault(require("./0.1.schema.json"));
exports.WrapManifestSchema_0_1 = _0_1_schema_json_1.default;
var WrapManifestVersions;
(function (WrapManifestVersions) {
    // NOTE: Patch fix for backwards compatability
    WrapManifestVersions["v0.1.0"] = "0.1.0";
    WrapManifestVersions["v0.1"] = "0.1";
})(WrapManifestVersions = exports.WrapManifestVersions || (exports.WrapManifestVersions = {}));
;
var WrapManifestAbiVersions;
(function (WrapManifestAbiVersions) {
    WrapManifestAbiVersions["v0.1"] = "0.1";
})(WrapManifestAbiVersions = exports.WrapManifestAbiVersions || (exports.WrapManifestAbiVersions = {}));
;
var WrapAbiVersions;
(function (WrapAbiVersions) {
    WrapAbiVersions["v0.1"] = "0.1";
})(WrapAbiVersions = exports.WrapAbiVersions || (exports.WrapAbiVersions = {}));
;
exports.latestWrapManifestVersion = WrapManifestVersions["v0.1"];
exports.latestWrapAbiVersion = "0.1";
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateWrapManifest", { enumerable: true, get: function () { return migrate_1.migrateWrapManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeWrapManifest", { enumerable: true, get: function () { return deserialize_1.deserializeWrapManifest; } });
var serialize_1 = require("./serialize");
Object.defineProperty(exports, "serializeWrapManifest", { enumerable: true, get: function () { return serialize_1.serializeWrapManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateWrapManifest", { enumerable: true, get: function () { return validate_1.validateWrapManifest; } });
