// @polywrap/tracing-js/build/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tracer = exports.TracingLevel = void 0;
var sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
var sdk_trace_web_1 = require("@opentelemetry/sdk-trace-web");
var api = __importStar(require("@opentelemetry/api"));
var resources_1 = require("@opentelemetry/resources");
var opentelemetry_console_exporter_1 = require("@fetsorn/opentelemetry-console-exporter");
var exporter_trace_otlp_http_1 = require("@opentelemetry/exporter-trace-otlp-http");
var TracingLevel;
(function (TracingLevel) {
    TracingLevel[TracingLevel["Low"] = 0] = "Low";
    TracingLevel[TracingLevel["Medium"] = 1] = "Medium";
    TracingLevel[TracingLevel["High"] = 2] = "High";
})(TracingLevel = exports.TracingLevel || (exports.TracingLevel = {}));
var isPromise = function (test) {
    return !!test && typeof test.then === "function";
};
var Tracer = /** @class */ (function () {
    function Tracer() {
    }
    Tracer.enableTracing = function (tracerName, config) {
        var _a, _b, _c, _d, _e;
        this._config.tracingLevel = (_a = config.tracingLevel) !== null && _a !== void 0 ? _a : this._config.tracingLevel;
        this._config.consoleEnabled = (_b = config.consoleEnabled) !== null && _b !== void 0 ? _b : this._config.consoleEnabled;
        this._config.consoleDetailed = (_c = config.consoleDetailed) !== null && _c !== void 0 ? _c : this._config.consoleDetailed;
        this._config.httpEnabled = (_d = config.httpEnabled) !== null && _d !== void 0 ? _d : this._config.httpEnabled;
        this._config.httpUrl = (_e = config.httpUrl) !== null && _e !== void 0 ? _e : this._config.httpUrl;
        this._initProvider(tracerName);
        if (this._provider) {
            this._tracer = this._provider.getTracer(tracerName);
        }
    };
    Tracer.disableTracing = function () {
        this._config.consoleEnabled = false;
        this._config.httpEnabled = false;
    };
    Tracer.startSpan = function (spanName, tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        if ((!this._config.consoleEnabled && !this._config.httpEnabled) ||
            tracingLevel < this._config.tracingLevel)
            return;
        var currentSpan = this._currentSpan();
        var span = this._tracer.startSpan(spanName, {}, currentSpan
            ? api.trace.setSpanContext(api.context.active(), currentSpan.spanContext())
            : undefined);
        this._pushSpan(span);
    };
    Tracer.endSpan = function (tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        if ((!this._config.consoleEnabled && !this._config.httpEnabled) ||
            tracingLevel < this._config.tracingLevel)
            return;
        var span = this._currentSpan();
        if (span) {
            span.end();
            this._popSpan();
        }
    };
    Tracer.setAttribute = function (attrName, data, tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        if ((!this._config.consoleEnabled && !this._config.httpEnabled) ||
            tracingLevel < this._config.tracingLevel)
            return;
        var span = this._currentSpan();
        if (span) {
            span.setAttribute(attrName, JSON.stringify(data));
        }
    };
    Tracer.addEvent = function (event, data, tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        if ((!this._config.consoleEnabled && !this._config.httpEnabled) ||
            tracingLevel < this._config.tracingLevel)
            return;
        var span = this._currentSpan();
        if (span) {
            span.addEvent(event, { data: JSON.stringify(data) });
        }
    };
    Tracer.recordException = function (error, tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        if ((!this._config.consoleEnabled && !this._config.httpEnabled) ||
            tracingLevel < this._config.tracingLevel)
            return;
        var span = this._currentSpan();
        if (span) {
            // recordException converts the error into a span event.
            span.recordException(error);
            // If the exception means the operation results in an
            // error state, you can also use it to update the span status.
            span.setStatus({ code: api.SpanStatusCode.ERROR });
        }
    };
    Tracer.traceMethod = function (span, tracingLevel) {
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        return function (target, key, descriptor) {
            var original = descriptor.value;
            descriptor.value = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                try {
                    Tracer.startSpan(span, tracingLevel);
                    Tracer.setAttribute("args", __assign({}, args), tracingLevel);
                    var result = original.apply(this, args);
                    if (isPromise(result)) {
                        return result.then(function (result) {
                            Tracer.setAttribute("output", result, tracingLevel);
                            Tracer.endSpan(tracingLevel);
                            return result;
                        });
                    }
                    else {
                        Tracer.setAttribute("output", result, tracingLevel);
                        Tracer.endSpan(tracingLevel);
                        return result;
                    }
                }
                catch (error) {
                    Tracer.recordException(error, tracingLevel);
                    Tracer.endSpan(tracingLevel);
                    throw error;
                }
            };
            return descriptor;
        };
    };
    Tracer.traceFunc = function (span, func, tracingLevel) {
        var _this = this;
        if (tracingLevel === void 0) { tracingLevel = TracingLevel.Low; }
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            try {
                _this.startSpan(span, tracingLevel);
                _this.setAttribute("args", __assign({}, args), tracingLevel);
                var result = func.apply(void 0, __spread(args));
                if (isPromise(result)) {
                    return result.then(function (result) {
                        _this.setAttribute("output", result, tracingLevel);
                        _this.endSpan(tracingLevel);
                        return result;
                    });
                }
                else {
                    _this.setAttribute("output", result, tracingLevel);
                    _this.endSpan(tracingLevel);
                    return result;
                }
            }
            catch (error) {
                _this.recordException(error, tracingLevel);
                _this.endSpan(tracingLevel);
                throw error;
            }
        };
    };
    Tracer._initProvider = function (name) {
        var providerResources = new resources_1.Resource({
            "service.name": name,
        });
        if (typeof window === "undefined") {
            this._provider = new sdk_trace_base_1.BasicTracerProvider({
                resource: providerResources,
            });
        }
        else {
            this._provider = new sdk_trace_web_1.WebTracerProvider({
                resource: providerResources,
            });
        }
        // Configure span processor to send spans to the exporter
        if (this._config.httpEnabled) {
            this._provider.addSpanProcessor(new sdk_trace_base_1.SimpleSpanProcessor(new exporter_trace_otlp_http_1.OTLPTraceExporter({ url: this._config.httpUrl })));
        }
        if (this._config.consoleEnabled) {
            this._provider.addSpanProcessor(new sdk_trace_base_1.BatchSpanProcessor(new opentelemetry_console_exporter_1.ConsoleExporterIcon({ isDetailed: this._config.consoleDetailed })));
        }
        this._provider.register();
    };
    Tracer._pushSpan = function (span) {
        this._spans.push(span);
    };
    Tracer._currentSpan = function () {
        return this._spans.slice(-1)[0];
    };
    Tracer._popSpan = function () {
        this._spans.pop();
    };
    Tracer._config = {
        consoleEnabled: false,
        tracingLevel: TracingLevel.High,
        consoleDetailed: false,
        httpEnabled: false,
        httpUrl: "http://localhost:4318/v1/traces",
    };
    Tracer._provider = null;
    Tracer._spans = [];
    return Tracer;
}());
exports.Tracer = Tracer;
