// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.infra/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateInfraManifest = exports.deserializeInfraManifest = exports.migrateInfraManifest = exports.latestInfraManifestFormat = exports.InfraManifestSchemaFiles = exports.InfraManifestFormats = void 0;
var InfraManifestFormats;
(function (InfraManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    InfraManifestFormats["v0.1"] = "0.1";
    InfraManifestFormats["v0.1.0"] = "0.1.0";
})(InfraManifestFormats = exports.InfraManifestFormats || (exports.InfraManifestFormats = {}));
exports.InfraManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.infra/0.1.0.json",
    "0.1.0": "formats/polywrap.infra/0.1.0.json",
};
exports.latestInfraManifestFormat = InfraManifestFormats["v0.1.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateInfraManifest", { enumerable: true, get: function () { return migrate_1.migrateInfraManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeInfraManifest", { enumerable: true, get: function () { return deserialize_1.deserializeInfraManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateInfraManifest", { enumerable: true, get: function () { return validate_1.validateInfraManifest; } });
