// @polywrap/uri-resolver-extensions-js/build/UriResolverWrapper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UriResolverWrapper = void 0;
var UriResolverExtensionFileReader_1 = require("./UriResolverExtensionFileReader");
var ResolverExtensionLoader_1 = require("./ResolverExtensionLoader");
var core_js_1 = require("@polywrap/core-js");
var result_1 = require("@polywrap/result");
var wasm_js_1 = require("@polywrap/wasm-js");
var uri_resolvers_js_1 = require("@polywrap/uri-resolvers-js");
var UriResolverWrapper = /** @class */ (function (_super) {
    __extends(UriResolverWrapper, _super);
    function UriResolverWrapper(implementationUri) {
        var _this = _super.call(this) || this;
        _this.implementationUri = implementationUri;
        _this.getStepDescription = function () {
            return "ResolverExtension (" + _this.implementationUri.uri + ")";
        };
        return _this;
    }
    UriResolverWrapper.prototype._tryResolveUri = function (uri, client, resolutionContext) {
        return __awaiter(this, void 0, void 0, function () {
            var result, uriOrManifest, wrapPackage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, tryResolveUriWithImplementation(uri, this.implementationUri, client, resolutionContext)];
                    case 1:
                        result = _a.sent();
                        if (!result.ok) {
                            return [2 /*return*/, core_js_1.UriResolutionResult.err(result.error)];
                        }
                        uriOrManifest = result.value;
                        if (uriOrManifest === null || uriOrManifest === void 0 ? void 0 : uriOrManifest.uri) {
                            return [2 /*return*/, core_js_1.UriResolutionResult.ok(new core_js_1.Uri(uriOrManifest.uri))];
                        }
                        else if (uriOrManifest === null || uriOrManifest === void 0 ? void 0 : uriOrManifest.manifest) {
                            wrapPackage = wasm_js_1.WasmPackage.from(uriOrManifest.manifest, new UriResolverExtensionFileReader_1.UriResolverExtensionFileReader(this.implementationUri, uri, client));
                            return [2 /*return*/, core_js_1.UriResolutionResult.ok(uri, wrapPackage)];
                        }
                        return [2 /*return*/, core_js_1.UriResolutionResult.ok(uri)];
                }
            });
        });
    };
    return UriResolverWrapper;
}(uri_resolvers_js_1.ResolverWithHistory));
exports.UriResolverWrapper = UriResolverWrapper;
var tryResolveUriWithImplementation = function (uri, implementationUri, client, resolutionContext) { return __awaiter(void 0, void 0, void 0, function () {
    var subContext, result, extensionWrapper, env, invokeResult, uriOrManifest;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                subContext = resolutionContext.createSubContext();
                return [4 /*yield*/, ResolverExtensionLoader_1.loadResolverExtension(uri, implementationUri, client, subContext)];
            case 1:
                result = _a.sent();
                if (!result.ok) {
                    return [2 /*return*/, result];
                }
                extensionWrapper = result.value;
                env = core_js_1.getEnvFromUriHistory(subContext.getResolutionPath(), client);
                return [4 /*yield*/, client.invokeWrapper({
                        wrapper: extensionWrapper,
                        uri: implementationUri.uri,
                        method: "tryResolveUri",
                        args: {
                            authority: uri.authority,
                            path: uri.path,
                        },
                        env: env === null || env === void 0 ? void 0 : env.env,
                    })];
            case 2:
                invokeResult = _a.sent();
                if (!invokeResult.ok) {
                    return [2 /*return*/, invokeResult];
                }
                uriOrManifest = invokeResult.value;
                return [2 /*return*/, result_1.ResultOk(uriOrManifest !== null && uriOrManifest !== void 0 ? uriOrManifest : undefined)];
        }
    });
}); };
