// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.meta/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateMetaManifest = exports.deserializeMetaManifest = exports.migrateMetaManifest = exports.latestMetaManifestFormat = exports.MetaManifestSchemaFiles = exports.MetaManifestFormats = void 0;
var MetaManifestFormats;
(function (MetaManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    MetaManifestFormats["v0.1"] = "0.1";
    MetaManifestFormats["v0.1.0"] = "0.1.0";
})(MetaManifestFormats = exports.MetaManifestFormats || (exports.MetaManifestFormats = {}));
exports.MetaManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.meta/0.1.0.json",
    "0.1.0": "formats/polywrap.meta/0.1.0.json",
};
exports.latestMetaManifestFormat = MetaManifestFormats["v0.1.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateMetaManifest", { enumerable: true, get: function () { return migrate_1.migrateMetaManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeMetaManifest", { enumerable: true, get: function () { return deserialize_1.deserializeMetaManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateMetaManifest", { enumerable: true, get: function () { return validate_1.validateMetaManifest; } });
