// polywrap/build/lib/logging/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
var logging_js_1 = require("@polywrap/logging-js");
var Logger = /** @class */ (function () {
    function Logger(logs) {
        this._logs = logs;
    }
    Logger.prototype.getLog = function (name) {
        return this._logs[name];
    };
    Logger.prototype.setLog = function (name, log) {
        this._logs[name] = log;
    };
    Logger.prototype.removeLog = function (name) {
        delete this._logs[name];
    };
    Logger.prototype.log = function (message, level) {
        Object.values(this._logs).forEach(function (log) { return log.log(message, level); });
    };
    Logger.prototype.debug = function (message) {
        this.log(message, logging_js_1.LogLevel.DEBUG);
    };
    Logger.prototype.info = function (message) {
        this.log(message, logging_js_1.LogLevel.INFO);
    };
    Logger.prototype.warn = function (message) {
        this.log(message, logging_js_1.LogLevel.WARN);
    };
    Logger.prototype.error = function (message) {
        this.log(message, logging_js_1.LogLevel.ERROR);
    };
    return Logger;
}());
exports.Logger = Logger;
