// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.deploy/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateDeployManifest = exports.deserializeDeployManifest = exports.migrateDeployManifest = exports.latestDeployManifestFormat = exports.DeployManifestSchemaFiles = exports.DeployManifestFormats = void 0;
var DeployManifestFormats;
(function (DeployManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    DeployManifestFormats["v0.1"] = "0.1";
    DeployManifestFormats["v0.1.0"] = "0.1.0";
    DeployManifestFormats["v0.2.0"] = "0.2.0";
})(DeployManifestFormats = exports.DeployManifestFormats || (exports.DeployManifestFormats = {}));
exports.DeployManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.deploy/0.1.0.json",
    "0.1.0": "formats/polywrap.deploy/0.1.0.json",
    "0.2.0": "formats/polywrap.deploy/0.2.0.json",
};
exports.latestDeployManifestFormat = DeployManifestFormats["v0.2.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateDeployManifest", { enumerable: true, get: function () { return migrate_1.migrateDeployManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeDeployManifest", { enumerable: true, get: function () { return deserialize_1.deserializeDeployManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateDeployManifest", { enumerable: true, get: function () { return validate_1.validateDeployManifest; } });
