// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.build/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateBuildManifest = exports.deserializeBuildManifest = exports.migrateBuildManifest = exports.latestBuildManifestFormat = exports.BuildManifestSchemaFiles = exports.BuildManifestFormats = void 0;
var BuildManifestFormats;
(function (BuildManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    BuildManifestFormats["v0.1"] = "0.1";
    BuildManifestFormats["v0.1.0"] = "0.1.0";
    BuildManifestFormats["v0.2.0"] = "0.2.0";
})(BuildManifestFormats = exports.BuildManifestFormats || (exports.BuildManifestFormats = {}));
exports.BuildManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.build/0.1.0.json",
    "0.1.0": "formats/polywrap.build/0.1.0.json",
    "0.2.0": "formats/polywrap.build/0.2.0.json",
};
exports.latestBuildManifestFormat = BuildManifestFormats["v0.2.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateBuildManifest", { enumerable: true, get: function () { return migrate_1.migrateBuildManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeBuildManifest", { enumerable: true, get: function () { return deserialize_1.deserializeBuildManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateBuildManifest", { enumerable: true, get: function () { return validate_1.validateBuildManifest; } });
