// @polywrap/logger-plugin-js/build/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.loggerPlugin = exports.LoggerPlugin = void 0;
var wrap_1 = require("./wrap");
var LoggerPlugin = /** @class */ (function (_super) {
    __extends(LoggerPlugin, _super);
    function LoggerPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LoggerPlugin.prototype.log = function (args) {
        if (this.config.logFunc) {
            return this.config.logFunc(args.level, args.message);
        }
        switch (args.level) {
            case "DEBUG":
            case wrap_1.Logger_LogLevelEnum.DEBUG:
                console.debug(args.message);
                break;
            case "WARN":
            case wrap_1.Logger_LogLevelEnum.WARN:
                console.warn(args.message);
                break;
            case "ERROR":
            case wrap_1.Logger_LogLevelEnum.ERROR:
                console.error(args.message);
                break;
            case "INFO":
            case wrap_1.Logger_LogLevelEnum.INFO:
                console.log(args.message);
                break;
            default:
                console.log(args.message);
        }
        return true;
    };
    return LoggerPlugin;
}(wrap_1.Module));
exports.LoggerPlugin = LoggerPlugin;
var loggerPlugin = function (config) {
    return {
        factory: function () { return new LoggerPlugin(config); },
        manifest: wrap_1.manifest,
    };
};
exports.loggerPlugin = loggerPlugin;
exports.plugin = exports.loggerPlugin;
