// polywrap/build/lib/docgen/schema/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBinding = exports.scriptPath = void 0;
var mustache_1 = __importDefault(require("mustache"));
var path_1 = __importDefault(require("path"));
var fs_1 = require("fs");
var schema_compose_1 = require("@polywrap/schema-compose");
exports.scriptPath = path_1.default.join(__dirname, "index.js");
var generateBinding = function (options) {
    var result = {
        output: {
            entries: [],
        },
        outputDirAbs: options.outputDirAbs,
    };
    var output = result.output;
    var renderTemplate = function (subPath, context, fileName) {
        var absPath = path_1.default.join(__dirname, subPath);
        var template = fs_1.readFileSync(absPath, { encoding: "utf-8" });
        output.entries.push({
            type: "File",
            name: fileName,
            data: mustache_1.default.render(template, context),
        });
    };
    // generate schema
    var schemaContext = {
        schema: schema_compose_1.renderSchema(options.abi, true),
    };
    renderTemplate("./templates/schema.mustache", schemaContext, "generated-schema.graphql");
    return result;
};
exports.generateBinding = generateBinding;
