// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.test/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePolywrapWorkflow = exports.deserializePolywrapWorkflow = exports.migratePolywrapWorkflow = exports.latestPolywrapWorkflowFormat = exports.PolywrapWorkflowSchemaFiles = exports.PolywrapWorkflowFormats = void 0;
var PolywrapWorkflowFormats;
(function (PolywrapWorkflowFormats) {
    // NOTE: Patch fix for backwards compatability
    PolywrapWorkflowFormats["v0.1"] = "0.1";
    PolywrapWorkflowFormats["v0.1.0"] = "0.1.0";
})(PolywrapWorkflowFormats = exports.PolywrapWorkflowFormats || (exports.PolywrapWorkflowFormats = {}));
exports.PolywrapWorkflowSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.test/0.1.0.json",
    "0.1.0": "formats/polywrap.test/0.1.0.json",
};
exports.latestPolywrapWorkflowFormat = PolywrapWorkflowFormats["v0.1.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migratePolywrapWorkflow", { enumerable: true, get: function () { return migrate_1.migratePolywrapWorkflow; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializePolywrapWorkflow", { enumerable: true, get: function () { return deserialize_1.deserializePolywrapWorkflow; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validatePolywrapWorkflow", { enumerable: true, get: function () { return validate_1.validatePolywrapWorkflow; } });
