// @polywrap/polywrap-manifest-types-js/build/formats/polywrap/migrators/0.1.0_to_0.2.0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrate = void 0;
function migrate(manifest, logger) {
    var _a, _b;
    var shouldHaveExtensions = manifest.build || manifest.deploy || manifest.meta;
    var maybeExtensions = {
        build: manifest.build,
        deploy: manifest.deploy,
        meta: manifest.meta,
    };
    if ((_a = manifest.import_redirects) === null || _a === void 0 ? void 0 : _a.some(function (x) {
        return x.schema.includes("build/schema.graphql");
    })) {
        logger === null || logger === void 0 ? void 0 : logger.warn("Detected a reference to \"build/schema.graphql\" in \"import_redirects\". Consider using \"build/wrap.info\" instead of \"build/schema.graphql\" in \"source.import_abis\", schema.graphql is no longer emitted as a build artifact.");
    }
    return {
        format: "0.2.0",
        project: {
            name: manifest.name,
            type: manifest.language,
        },
        source: {
            schema: manifest.schema,
            module: manifest.module,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            import_abis: (_b = manifest.import_redirects) === null || _b === void 0 ? void 0 : _b.map(function (x) { return ({
                uri: x.uri,
                abi: x.schema,
            }); }),
        },
        extensions: shouldHaveExtensions ? maybeExtensions : undefined,
        // eslint-disable-next-line @typescript-eslint/naming-convention
        __type: "PolywrapManifest",
    };
}
exports.migrate = migrate;
