// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.app/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAppManifest = exports.deserializeAppManifest = exports.migrateAppManifest = exports.latestAppManifestFormat = exports.AppManifestSchemaFiles = exports.AppManifestFormats = void 0;
var AppManifestFormats;
(function (AppManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    AppManifestFormats["v0.1"] = "0.1";
    AppManifestFormats["v0.1.0"] = "0.1.0";
    AppManifestFormats["v0.2.0"] = "0.2.0";
})(AppManifestFormats = exports.AppManifestFormats || (exports.AppManifestFormats = {}));
exports.AppManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.app/0.1.0.json",
    "0.1.0": "formats/polywrap.app/0.1.0.json",
    "0.2.0": "formats/polywrap.app/0.2.0.json",
};
exports.latestAppManifestFormat = AppManifestFormats["v0.2.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateAppManifest", { enumerable: true, get: function () { return migrate_1.migrateAppManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeAppManifest", { enumerable: true, get: function () { return deserialize_1.deserializeAppManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateAppManifest", { enumerable: true, get: function () { return validate_1.validateAppManifest; } });
